app/app.go

```diff
@@ -343,35 +343,54 @@ func (a *App) Start() error {
 		logger.Info("Actions Started")
 	}
 
+	lifecycleTriggers := make(map[string]*triggerWrapper)
+	normalTriggers := make(map[string]*triggerWrapper)
+
+	for id, trgW := range a.triggers {
+		if _, ok := trgW.trg.(LifecycleAware); ok {
+			lifecycleTriggers[id] = trgW
+		} else {
+			normalTriggers[id] = trgW
+		}
+	}
+
 	// Start the triggers
 	logger.Info("Starting Triggers...")
 
 	var failed []string
 
-	for id, trg := range a.triggers {
-		statusInfo := trg.status
-		err := managed.Start(fmt.Sprintf("Trigger [ %s ]", id), trg.trg)
+	// Start Lifecycle triggers
+	for id, trgW := range lifecycleTriggers {
+		ok, err := a.startTrigger(id, trgW)
 		if err != nil {
-			if a.stopOnError {
-				return fmt.Errorf("error starting Trigger[%s] : %s", id, err)
-			}
-			logger.Infof("Trigger [%s] failed to start due to error [%s]", id, err.Error())
-			statusInfo.Status = managed.StatusFailed
-			statusInfo.Error = err
-			logger.Debugf("StackTrace: %s", debug.Stack())
+			return err
+		}
+		if !ok {
+			failed = append(failed, id)
+		}
+	}
+
+	// Invoke OnStartup for lifecycle aware triggers
+	for _, trgW := range lifecycleTriggers {
+		if trgW.status.Status == managed.StatusStarted  {
+			lca, _ := trgW.trg.(LifecycleAware)
+			lca.OnStartup()
```

Comment:
I think we should handle errors which through from OnStartUp or OnShutdown.  either ignore or exist(controlled by stop on error env var)

Reply:
@mellistibco what do you think.  This is not the start/stop of a trigger.  This is more like you received a "start" event hand you kick off the corresponding handler/action.  The action has an error...     I guess it makes sense if this "startup" action is necessary.  Do we tie into the existing "stop on error"... or should this be separate ( almost of this startup action is not critical to startup)?

Reply:
@mellistibco @vijaynalawade Any thoughts?

Reply:
@fm-tibco @lixingwang : I feel we should simply let engine fail in case OnStartup action returns any error. This is like initializing variables. If something bad happens, runtime would fail.

Reply:
That sounds good, I'll have the interface return error and stop engine on failure.  We can always revisit the behavior on failure.